[FEATURE] Added integration test system

adirajagopal opened the pull request:
This isn't exactly as we had planned, but this is what seemed to work most reliably with least code repeats. Basically, this does require nc_pkg.py in any example to have an implementation for --build, --exec and --down. If it has these three, then we can run any test we want form the tests/integration directory.

adirajagopal committed:
[FEATURE] Added integration test system
- Updated basic_example nc_pkg.py to run with docker-compose
- Deleted publish.sh
- Added working integration test for basic_example
- Added integration test for nsys_example
- Organized file struture for integration tests
- Removed secrets from nsys-rep

diederik-vink commented on examples/basic_example/nc_pkg.py:
```diff
@@ -1,86 +1,253 @@
+    env['CURRENT_DIR'] = str(current_dir.absolute())
+    
+    # Set UID, GID, DISPLAY
+    env['UID'] = str(os.getuid())
+    env['GID'] = str(os.getgid())
+    env['DISPLAY'] = os.environ.get('DISPLAY', ':0')
```
Are we sure every machine has these?

adirajagopal replied:
Si

vineye25 commented on examples/basic_example/.gitignore:
[QUESTION] Is this the right way to treat the LFS files?

I was asking ChatGPT and it was saying we shouldn't be seeing these in the line change diffs(`+326k`), but I'm not so familiar with Git LFS

adirajagopal replied:
Updated git lfs tracking. Files weren't being tracked properly after some file structure changes.

diederik-vink commented on examples/basic_example/nc_pkg.py:
```diff
@@ -1,86 +1,253 @@
+    
+    # Start the container
+    print("Starting container...")
+    run_compose_command(compose_files, ["up", "-d"], env, capture_output=False).check_returncode()
```
THis does not ensure anything is running, this forces a restart. Should be renamed or change functionality

adirajagopal replied:
Updated naming

diederik-vink commented on tests/integration/test_basic_example/test_basic_example.py:
```diff
@@ -0,0 +1,63 @@
+        )
+
+@pytest.mark.integration
+def test_basic_example_link_only(example_dir: Path, 
+                                 symlink_sdk,
+                                 build_and_kill_basic_example_docker_container,
+                                 copy_basic_example_gold_result,
+                                 test_dir: Path):
+    print("Running Test ...")
+    test_and_check_cmd = f"""
+    python main.py --link-only test_files/test_trace.pt.trace.json &&
+    mv test_files/test_trace.linked.pt.trace.json .pytest/ &&
+    diff .pytest/test_trace.linked.pt.trace.json .pytest/gold.pt.trace.json
+    """
+    subprocess.run(
+        [sys.executable, "nc_pkg.py", "--exec", test_and_check_cmd],
+        cwd = example_dir,
+        check=True
+    )
```
is tests/integration/compare_files.py not needed to perform this diff?

diederik-vink commented on tests/integration/test_nsys_example/test_nsys_example.py:
```diff
@@ -0,0 +1,58 @@
+        )
+
+@pytest.mark.integration
+def test_nsys_example(example_dir: Path, 
+                     symlink_sdk,
+                     build_and_kill_nsys_example_docker_container,
+                     copy_nsys_example_gold_result,
+                     test_dir: Path):
+    print("Running Test ...")
+    test_and_check_cmd = f"""
+    python convert_nsys.py --input test_files/test_trace.nsys-rep &&
+    test -f test_trace.sqlite &&
+    mv test_trace.json test_trace.sqlite .pytest/ &&
+    diff .pytest/test_trace.json .pytest/gold_nsys_example.json
+    """
+    subprocess.run(
+        [sys.executable, "nc_pkg.py", "--exec", test_and_check_cmd],
+        cwd = example_dir,
+        check=True
+    )
```
same here about using tests/integration/compare_files.py as the diff runner?

diederik-vink commented on tests/integration/utils.py:
```diff
@@ -0,0 +1,195 @@
+    Returns:
+        Dictionary of environment variables
+    """
+    env = os.environ.copy()
```
why are we doing this?

adirajagopal replied:
Updated to just have this start as {}, not os.environ.copy()

vineye25 commented on tests/integration/compare_files.py:
```diff
@@ -0,0 +1,161 @@
+    - File paths that may differ
+    - Other non-deterministic fields
+    """
+    if isinstance(data, dict):
+        normalized = {}
+        for key, value in data.items():
+            # Skip or normalize non-deterministic fields
+            if key in ["ts", "timestamp", "time", "startTime", "endTime"]:
+                # Keep structure but normalize timestamp to 0 for comparison
+                normalized[key] = 0
+            elif key in ["id", "uuid", "pid", "tid"]:
+                # Keep structure but normalize IDs
+                normalized[key] = 0
+            elif key in ["name", "cat"]:
+                # Keep these as they're important for comparison
+                normalized[key] = value
+            elif isinstance(value, (dict, list)):
+                normalized[key] = normalize_json_data(value)
+            else:
+                normalized[key] = value
+        return normalized
```
Is it ok to strip out `pid` and `tid`?

But also isn't this if/elif/else statement actually just doing:
```python
if key in NON_DETERMINISTIC_FIELDS:
    normalized[key] = 0
elif isinstance(value, (dict, list)):
    normalized[key] = normalize_json_data(value)
else:
    normalized[key] = value
```

adirajagopal replied:
Removing compare_files for now.

adirajagopal committed:
[PR-FIX] https://github.com/nCompass-tech/ncompass/pull/17
- Update git lfs tracking of large files
- Renamed ensure_running to force_restart
- Docker compose setup doesn't capture entire env
